query/query_proto.go

```diff
@@ -0,0 +1,413 @@
+package query
+
+import (
+	"fmt"
+	"regexp/syntax"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/grafana/regexp"
+	v1 "github.com/sourcegraph/zoekt/grpc/v1"
+)
+
+func QToProto(q Q) *v1.Q {
+	switch v := q.(type) {
+	case RawConfig:
+		return &v1.Q{Query: &v1.Q_RawConfig{RawConfig: v.ToProto()}}
+	case *Regexp:
+		return &v1.Q{Query: &v1.Q_Regexp{Regexp: v.ToProto()}}
+	case *Symbol:
+		return &v1.Q{Query: &v1.Q_Symbol{Symbol: v.ToProto()}}
+	case *Language:
+		return &v1.Q{Query: &v1.Q_Language{Language: v.ToProto()}}
+	case *Const:
+		return &v1.Q{Query: &v1.Q_Const{Const: v.ToProto()}}
+	case *Repo:
+		return &v1.Q{Query: &v1.Q_Repo{Repo: v.ToProto()}}
+	case *RepoRegexp:
+		return &v1.Q{Query: &v1.Q_RepoRegexp{RepoRegexp: v.ToProto()}}
+	case *BranchesRepos:
+		return &v1.Q{Query: &v1.Q_BranchesRepos{BranchesRepos: v.ToProto()}}
+	case *RepoIDs:
+		return &v1.Q{Query: &v1.Q_RepoIds{RepoIds: v.ToProto()}}
+	case *RepoSet:
+		return &v1.Q{Query: &v1.Q_RepoSet{RepoSet: v.ToProto()}}
+	case *FileNameSet:
+		return &v1.Q{Query: &v1.Q_FileNameSet{FileNameSet: v.ToProto()}}
+	case *Type:
+		return &v1.Q{Query: &v1.Q_Type{Type: v.ToProto()}}
+	case *Substring:
+		return &v1.Q{Query: &v1.Q_Substring{Substring: v.ToProto()}}
+	case *And:
+		return &v1.Q{Query: &v1.Q_And{And: v.ToProto()}}
+	case *Or:
+		return &v1.Q{Query: &v1.Q_Or{Or: v.ToProto()}}
+	case *Not:
+		return &v1.Q{Query: &v1.Q_Not{Not: v.ToProto()}}
+	case *Branch:
+		return &v1.Q{Query: &v1.Q_Branch{Branch: v.ToProto()}}
+	default:
+		// The following nodes do not have a proto representation:
+		// - GobCache: only needed for Gob encoding
```

Comment:
how are we handling what GobCache did? IE how do we pay the cost of serialisation only once in sourcegraph-frontend, rather than serialising per replica.

Reply:
That looks like it should work, just need to avoid shooting ourselves in the foot with maybe some extra checks / maybe using a oneof? IE we only really care about paying the cost of Q once.

Taking a step back, we may actually be in a position these days were the size of the queries are no longer huge anymore? Would be interesting to add some instrumentation here (if it doesn't already exist) and confirm if this hack is still needed.

Reply:
We log the query size in bytes on the trace, so I just did a little bit of manual testing on s2, and I can't get the query size to exceed ~1.5K. This seems cheap enough that we don't need to worry about this anymore, but I'll monitor on dotcom as well once tracing is fixed